---
title: "Probe Mapping"
author: "Kate (kshutta@hsph.harvard.edu)"
date: "3/25/2022"
output: html_document
---

```{R setup}
library(aws.s3)
library(data.table)
library(tidyverse)
# need: https://zwdzwd.github.io/InfiniumAnnotation
```

The purpose of this code is map probes to gene, going from a matrix of probes x samples to a matrix of genes x samples. At this state, we want to simply map each promoter to its nearest gene. For this we need the probe manifest. See this helpful file: https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/methylationepic/infinium-methylationepic-manifest-column-headings.pdf

```{R}
# source hg38 with gencode 36 from https://zwdzwd.github.io/InfiniumAnnotation

download.file('https://zhouserver.research.chop.edu/InfiniumAnnotation/20210615/HM450/HM450.hg38.manifest.gencode.v36.tsv.gz', destfile = "../../data/external/HM450.hg38.manifest.gencode.v36.tsv.gz")

# unzip
system2(command="gunzip",args=c("../../data/external/HM450.hg38.manifest.gencode.v36.tsv.gz"))

# load into memory
manifest = data.frame(fread("../../data/external/HM450.hg38.manifest.gencode.v36.tsv",sep="\t",header=T))
```

```{R}
names(manifest)
```

```{R}
manifest$probeID[1]
manifest$genesUniq[1]
manifest$geneNames[1]
manifest$distToTSS[1]
```

What about probes that map to multiple genes?


```{R}
manifest$probeID[2]
manifest$genesUniq[2]
manifest$geneNames[2]
manifest$distToTSS[2]
```

Order of operations: for each probe, need to (1) see if there is >1 gene within TSS-1500 (2) if so, see if they all correspond to the same gene (splice isoforms?), (3) if so, label probe with gene, (4) if not, map to the closer gene. Note this is arbitrary choice!! May make sense to allow to map to both genes?

```{R}
processRow = function(x) # x is one row of the manifest
{
  x = as.data.frame(x)
  genes = str_split(x$geneNames,";",simplify=T)
  tssDist = as.numeric(str_split(x$distToTSS,";",simplify=T))
  promoterRegions = which(tssDist > -1500 & tssDist < 0) # negative = upstream of TSS
  promoterMap = genes[promoterRegions]
  if(length(unique(promoterMap)) > 1)
  {
    index = which.max(tssDist[promoterRegions])
    thisRow = c(x$probeID,promoterMap[index],tssDist[promoterRegions][index])
    return(thisRow)
  }
  
  thisGene = ifelse(length(promoterMap)>0,unique(promoterMap), NA)
  thisDist = ifelse(length(promoterMap)>0,max(tssDist[promoterRegions]),NA) 
  return(c(x$probeID,thisGene,thisDist))
}
```

 Perform the mapping:

```{R}
mymap = matrix(rep(NA,3*nrow(manifest)),ncol=3)
for(i in 1:nrow(manifest))
{
  if(i %% 10000 == 0) print(paste("probe number:",i))
  mymap[i,]=processRow(manifest[i,])
}
```

confirm that nothing is mapped to multiple genes (sanity check)

```{R}
grep(";",mymap[,2])
```
Write table for those probes that mapped to promoter regions:

```{R}
write.table(na.omit(mymap),file="../../data/processed/PromoterProbeMap450K.txt",sep="\t",row.names=F,quote=F)
```

Upload this map to S3 for reproducibility:

```{R}
Sys.setenv("AWS_PROFILE" = "MFA")
put_object(file="../../data/processed/PromoterProbeMap450K.txt",
      object = "PromoterProbeMap450K.txt",
      bucket = "netzoo/supData/dragon",
      region="us-east-2",
      multipart=F)
```

Here is code to make a second version in which everything with multiple genes is split to long format. This was needed when we allowed a probe to map to multiple genes, keeping it here in case we decide to do that later. 

```{R eval=F}
mymap = read.table("../../data/processed/PromoterProbeMap450K.txt",sep="\t",header=T)
names(mymap) = c("probeID","gene")
doubleGenes = mymap[grep(";",mymap[,2]),]
for(i in 1:nrow(doubleGenes))
{
  if(i %% 1000 == 0) print(i)
  theseSplitGenes = unique(str_split(doubleGenes[i,2],";",simplify=T)[1,])
  miniDF = cbind.data.frame(doubleGenes[i,1],theseSplitGenes) 
  names(miniDF)[1]="probeID"
  longForm = pivot_longer(miniDF,cols=2:ncol(miniDF),names_to=NULL,values_to = "gene")
  mymap = rbind.data.frame(mymap,longForm)
}

mymap_long = mymap[-grep(";",mymap[,2]),]
write.table(na.omit(mymap_long),file="../../data/processed/ProbeMap450K_longForm.txt",sep="\t",row.names=F,quote=F)
```
